### PR TITLE
read composite on root path

### PR DESCRIPTION
**Pull request by mgdlkundera**
Integration tests for root

### CODE REVIEW

**Comment by sbernard31** [comment]
(@mgdlkundera  let me know when this is in a reviewable state)

**Comment by mgdlkundera** [comment]
@sbernard31 It's ready for review

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/LwM2mRoot.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public int getId() {
+        return 0;
+    }
+
+    @Override
+    public void accept(LwM2mNodeVisitor visitor) {
+        // TODO: visitors currently don't support a visit from LwM2mRoot
+        visitor.visit(this);
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/LwM2mRoot.java:47]: I'm not sure to get this TODO ?

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/LwM2mRoot.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public String toPrettyString(LwM2mPath path) {
+        return null;
+    }
+
+    @Override
+    public StringBuilder appendPrettyNode(StringBuilder b, LwM2mPath path) {
+        return null;
+    }
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/LwM2mRoot.java:58]: I think this should be implemented.

**Comment by sbernard31** [comment]
Not a big deal:slightly_smiling_face:   but maybe you could find : 
- a better PR name
- a better description or at least point to corresponding issue. (someone who clicks on the PR should vaguely understand what it is about)

**Comment by sbernard31** [comment]
(let me know once, you think the code must be reviewed again, no urgency on my side take, do not hesitate to take time to review your own code like if you review someone else code)

**Comment by sbernard31** [comment]
I will be unavailable next few days. I'll be back on Thursday November 2nd.

When I return, I plan to work on this.

**Comment by mgdlkundera** [comment]
Thank you. I will be out of the office until Monday November 6nd.

What do you think about possibility of releasing M14 version containing this issue around November 23?

**Comment by sbernard31** [comment]
> What do you think about possibility of releasing M14 version containing this issue around November 23?

We can try, I will create the corresponding issue.

**Comment by mgdlkundera** [comment]
I have one more question, can I add to this pull request observe for root?

**Comment by sbernard31** [comment]
@mgdlkundera I'm working on this PR, I hope I will provide something before the end of the day. So I don't advice to work more on it for now.

About observe, I advice to start to work on it in another PR.  :slightly_smiling_face:

**Comment by sbernard31** [comment]
I created a PR based on this one : #1534

I added a new Interface `LwM2mChildNode` because until now a `LwM2mNode` had a `getId()` method but for LwM2mRoot this is not true. 

@mgdlkundera, @JaroslawLegierski let me know if you plan to review the PR. :pray:

**Comment by sbernard31** [comment]
@mgdlkundera, @JaroslawLegierski  please let me know if you plan to review because if you don't plan better to put this in `master` so we can have some beta tester with sandbox users.

**Comment by JaroslawLegierski** [comment]
> @mgdlkundera, @JaroslawLegierski please let me know if you plan to review because if you don't plan to do it better to put this in `master` so we can have some beta tester with sandbox users.

If it is urgent then I suggest integrating it with the master (I can review it on Tuesday next week at the earliest).

**Comment by sbernard31** [comment]
> If it is urgent then I suggest integrating it with the master (I can review it on Tuesday next week at the earliest).

On my side there is no urgency. 
But If we want to target 23 November for 2.0.0-M14(https://github.com/eclipse-leshan/leshan/issues/1531), I think it makes sense to put it in `master` as soon as possible to see possible regression from sandbox users.

Looking at : https://github.com/zephyrproject-rtos/zephyr/pull/64013
I think I will try to add at least REST API support to **leshan-server-demo**, so zephyr could test this news feature.

**Comment by sbernard31** [comment]
@JaroslawLegierski , @mgdlkundera, 

As planned, I integrated https://github.com/eclipse-leshan/leshan/pull/1534 in `master`
So we will get maybe get some users sandboxes feedback (or at least detect possible regression)
But this is still possible to you to review the code, if there is something not good we can still fix/adapt it in another PR.

Note that I also added support in **leshan-server-demo**.

Thx both of you for you work/help on this :pray: !

**Comment by mgdlkundera** [comment]
I prepare implementation of observe for root, here is a link:
https://github.com/eclipse-leshan/leshan/compare/master...JaroslawLegierski:leshan:opl/observe_for_root

Can I create new pull request from this branch?

**Comment by sbernard31** [comment]
Yep.
Please could you provide code much more similar to Read-Composite and could you also add integration tests.

**Comment by mgdlkundera** [comment]
I improved code and add test. Here is pull request:
https://github.com/eclipse-leshan/leshan/pull/1539

**Pull request closed without merging**